data_diff/diff_tables.py

```diff
@@ -208,6 +208,10 @@ def _diff_tables_wrapper(self, table1: TableSegment, table2: TableSegment, info_
             event_json = create_start_event_json(options)
             run_as_daemon(send_event_json, event_json)
 
+        if table1.database.dialect.PREVENT_OVERFLOW_WHEN_CONCAT or table2.database.dialect.PREVENT_OVERFLOW_WHEN_CONCAT:
+            table1.database.dialect.enable_preventing_type_overflow()
+            table2.database.dialect.enable_preventing_type_overflow()
```

Comment:
❗️This would be problematic. 

Sadly, at this moment, dialect is a global class-level object. And we reuse it for multiple runs across multiple databases. If it is mutated here for a pair of databases X-Y, it will also affect other pairs X-Z & Y-Z by breaking them — because the upper case constants are not modified, only the lower-case fields are, so the hashing logic will be different.

Restoring the fields to their original values will also not be reliable, as some xdb diffs can be prematurely terminated by exceptions and the workers reused after that for other xdb diffs — catching the issues with the modified dialect.

A good solution would be to make `database.dialect` a property that is created per database object and discarded afterwards. This is on the roadmap anyway for different reasons, but not a priority. 

Though it seems as easy as doing:

```python
@attrs.define
class WhateverDatabase:
    dialect: BaseDialect = attrs.field(factory=WhateverDialect)
```

Instead of the current

```python
@attrs.define
class WhateverDatabase:
    dialect = WhateverDialect()
```

Can you please try it?

A less good way would be to store the state somewhere else, e.g. in databases or in ColTypes (created for each run and discarded afterwards).

Reply:
It is a good catch, thank you very much. I just thought that `attrs.define` makes as the same thing as dataclass but we still have a classvar here. I will try to fix it using your suggestion.

Reply:
@nolar your solution does not work.

Reply:
attrs works similar to dataclasses: if you want a classvar, annotate its type as `varname: ClassVar[BaseDialect] = ...` — in that case, attrs will ignore it entirely. The same as when there is no type annotation for a field, but that is not nice.

>  Cons here are it is needed to create it each time when dialect is accessed and implement the property for each connector.

In that case, you will lose its state, including the overflow prevention flag — the `db.dialect` is accessed many times at different occasions.

> your solution does not work.

What is the exact problem? Error?

Reply:
> In that case, you will lose its state, including the overflow prevention flag — the db.dialect is accessed many times at different occasions.

Yes, seems true

> What is the exact problem? Error?

For example, if i run it postgres vs postgres, the error is
`AttributeError: 'PostgreSQL' object has no attribute 'dialect'`

Reply:
I managed to emulate it in a script:
```
import attrs
import abc


@attrs.define(frozen=False)
class BaseParam:
    _flag_enabled: bool = False

    def enable(self):
        self._flag_enabled = True

    def print(self):
        print('base param')


@attrs.define(frozen=False)
class Param(BaseParam):
    def print(self):
        print('param')


@attrs.define(frozen=False, kw_only=True)
class Base(abc.ABC):
    var: int = 1

    @property
    @abc.abstractmethod
    def param(self) -> BaseParam:
        pass


@attrs.define(frozen=False, kw_only=True)
class MyClass(Base):
    param: BaseParam = attrs.field(factory=Param)
    _args: dict[str, any]

    def __init__(self, *, var, **kw):
        super().__init__(var=var)
        self._args = kw


if __name__ == '__main__':
    kw = {'a': 1, 'b': 2}
    c1 = MyClass(var=1, **kw)
    c2 = MyClass(var=1, **kw)
    c1.param.print()
```

Reply:
@nolar 
i managed to rewrite it to use the local instance of the dialect class instead of class var and pushed it. The short idea in the script below:
```
import attrs
import abc


@attrs.define(frozen=False)
class BaseParam:
    _flag_enabled: bool = False

    def enable(self):
        self._flag_enabled = True

    def print(self):
        print('base param')


@attrs.define(frozen=False)
class Param(BaseParam):
    def print(self):
        print('param')


@attrs.define(frozen=False, kw_only=True)
class Base(abc.ABC):
    Class = BaseParam
    var: int = 1
    _param: BaseParam = None

    @property
    def param(self) -> BaseParam:
        if not self._param:
            self._param = self.Class()
        return self._param


@attrs.define(frozen=False, kw_only=True)
class MyClass(Base):
    Class = Param
    _args: dict[str, any]

    def __init__(self, *, var, **kw):
        super().__init__(var=var)
        self._args = kw


if __name__ == '__main__':
    kw = {'a': 1, 'b': 2}
    c1 = MyClass(var=1, **kw)
    c2 = MyClass(var=1, **kw)
    c1.param.print()
    print(c1._param)
```